File: oauth2library/src/main/java/ca/mimic/oauth2library/AuthState.java
package ca.mimic.oauth2library;

class AuthState {
    static final int ACCESS_TOKEN = 0;
    static final int REFRESH_TOKEN = 1;

    private static final int NO_AUTH = 0;
    private static final int BASIC_AUTH = 1;
    private static final int AUTHORIZATION_AUTH = 2;
    private static final int FINAL_AUTH = 3;

    private static final int[] ACCESS_STATES = new int[]{
            NO_AUTH,
            BASIC_AUTH,
            AUTHORIZATION_AUTH,
            FINAL_AUTH
    };

    private static final int[] REFRESH_STATES = new int[]{
            NO_AUTH,
            AUTHORIZATION_AUTH,
            FINAL_AUTH
    };

    private int[] state;

    private int tokenType;
    private int position;
    private boolean initialAuth;

    AuthState(int tokenType) {
        this.tokenType = tokenType;

        switch (tokenType) {
            case ACCESS_TOKEN:
                state = ACCESS_STATES;
                break;
            case REFRESH_TOKEN:
                state = REFRESH_STATES;
                break;
        }
    }

    void nextState() {
        position++;
    }

    boolean isFinalAuth() {
        return state[position] == FINAL_AUTH;
    }

    boolean isBasicAuth() {
        return state[position] == BASIC_AUTH;
    }

    boolean isAuthorizationAuth() {
        return state[position] == AUTHORIZATION_AUTH;
    }

    public int getTokenType() {
        return tokenType;
    }
}

File: oauth2library/src/main/java/com/pillohealth/oauth2library/AuthState.kt
package com.pillohealth.oauth2library

internal class AuthState(tokenType: Int) {

    private var state: IntArray? = null

    private var position: Int = 0

    val isFinalAuth: Boolean
        get() = state!![position] == FINAL_AUTH

    val isBasicAuth: Boolean
        get() = state!![position] == BASIC_AUTH

    val isAuthorizationAuth: Boolean
        get() = state!![position] == AUTHORIZATION_AUTH

    init {
        when (tokenType) {
            ACCESS_TOKEN -> state = ACCESS_STATES
            REFRESH_TOKEN -> state = REFRESH_STATES
            else -> state = ACCESS_STATES
        }
    }

    fun nextState() {
        position++
    }

    companion object {
        internal val ACCESS_TOKEN = 0
        internal val REFRESH_TOKEN = 1

        private val NO_AUTH = 0
        private val BASIC_AUTH = 1
        private val AUTHORIZATION_AUTH = 2
        private val FINAL_AUTH = 3

        private val ACCESS_STATES = intArrayOf(NO_AUTH, BASIC_AUTH, AUTHORIZATION_AUTH, FINAL_AUTH)

        private val REFRESH_STATES = intArrayOf(NO_AUTH, AUTHORIZATION_AUTH, FINAL_AUTH)
    }
}
